### DISCUSSION TITLE
Count CG motifs in Roslin genome

### DISCUSSION THREAD

**Discussion by yaaminiv** [RobertsLab/resources#1208] [Q&A] [answered]
I want to count the number of CG motifs in the Roslin genome. I used the following command to do so:

![Screen Shot 2021-05-11 at 10 52 32 AM](https://user-images.githubusercontent.com/22335838/117862022-fe5e9b80-b246-11eb-81ff-9c8b2c7eb091.png)

~4 million CpGs seems low (*C. virginica* has ~14 million). When I look at `methylKit::unite` output, I get ~4.5 million CpGs with data. Clearly I'm counting the CGs incorrectly. What am I missing?

*edited to include `unite` information`*

**Comment by kubu4** [comment]
This will miss instances when a CG occurs at a newline.

e.g.:

```
AATCCGATAGC
GATTTACTATCG
```

**Reply by yaaminiv** [reply]
Ah! Looked into my old code and am using `fgrep -o -i CG {genome-file} | wc -l` instead!

**Reply by kubu4** [reply]
I don't think that will count across newlines, though...

**Reply by yaaminiv** [reply]
It might? I got ~13 million CG motifs. I will try the tab-delimited approach and compare!

**Reply by yaaminiv** [reply]
For reference, here's the issue where we had a discussion about `fgrep`: https://github.com/RobertsLab/resources/issues/685#issuecomment-488439875

**Reply by kubu4** [reply]
Ah! Funny we've discussed this before!

**Comment by kubu4** [comment]
I think the approach is to convert the FastA into a tab-delimited file. I use seqkit to do this, usually. Then, count with awk (possibly need to use gawk, to accommodate large number of fields). E.g.

```shell
  # Use seqkit to convert FastA to tab-delimited and print sequence length 
  ${seqkit} fx2tab \
  --length \
  ${fasta} \
  > ${fasta}.tab

  # Print only sequences to new file
  gawk '{ print $2 }' ${fn}_tab > ${fasta}.tab2
  
  # Delimit sequences on CGs and print the number of fields minus 1 to get the number of CGs present.
  gawk -F\[Cc][Gg] '{print NF-1}' ${fasta}.tab2 > CG

**Reply by yaaminiv** [reply]
Here's the code I'm using (I modified the last chunk to count the number of CGs instead of printing how many there were on each line:

![Screen Shot 2021-05-11 at 11 55 52 AM](https://user-images.githubusercontent.com/22335838/117869590-d7589780-b24f-11eb-9b70-b7c9f1db0755.png)

I get 0 CGs. I think this is because the `seqkit fx2tab` is not working:

![Screen Shot 2021-05-11 at 11 54 25 AM](https://user-images.githubusercontent.com/22335838/117869431-a4160880-b24f-11eb-8939-16e7f503f03c.png)

How should I modify my code?

**Reply by sr320** [reply]
You could convert using this instead
```

#Converting FASTA to tabular format and placing output file in analyses directory
!perl -e '$count=0; $len=0; while(<>) {s/\r?\n//; s/\t/ /g; if (s/^>//) { if ($. != 1) {print "\n"} s/ |$/\t/; $count++; $_ .= "\t";} else {s/ //g; $len += length($_)} print $_;} print "\n"; warn "\nConverted $count FASTA records in $. lines to tabular format\nTotal sequence length: $len\n\n";' \
33496_Ahyacinthus_CoralContigs.fasta > ../../analyses/Ahya/fasta2tab
```

**Reply by yaaminiv** [reply]
![Screen Shot 2021-05-11 at 12 04 03 PM](https://user-images.githubusercontent.com/22335838/117870452-015e8980-b251-11eb-952b-8baa6c3c8814.png)

Still running into an error with conversion to tab...when I look at `head` it prints the entire file

**Reply by yaaminiv** [reply]
I might abandon this since I just wanted a rough count of the CGs in the genome. I will make a CG motif track for IGV as well, which will then get me an exact count!

**Comment by yaaminiv** [comment]
@sr320 How do you usually generate the CG motif files for genomes? Perhaps that's what I should be doing, so I have that as an IGV track

**Reply by sr320** [reply]
normally EMBOSS Fuzznuc.... 
though have used could Sam referenced.. another eg https://github.com/jldimond/Coral-CpG/blob/master/ipynb/Ahya_CpG_ratio.ipynb

**Reply by yaaminiv** [reply]
Could you share your EMBOSS Fuzznuc code? I'd like to generate CG track for the genome. Beyond counting CGs, having the separate track would be helpful.

**Reply by sr320** [reply]
no EMBOSS code - used Galaxy

**Reply by yaaminiv** [reply]
Does this look right? 

![Screen Shot 2021-05-11 at 12 19 30 PM](https://user-images.githubusercontent.com/22335838/117872110-2522cf00-b253-11eb-8766-a5cb7c15515d.png)

**Reply by sr320** [reply]
Yes (I have it running to now on Galaxy) hopefully we get same product

**Reply by sr320** [reply]
Here is a link to the GFF https://usegalaxy.org/datasets/bbd44e69cb8906b5bba9b46c41a81f2b/display?to_ext=gff

**Reply by kubu4** [reply]
Here's fuzznuc code I've used in the past:

https://robertslab.github.io/sams-notebook/2019/08/21/Data-Wrangling-Create-a-CpG-GFF-from-Pgenerosa_v074-using-EMBOSS-fuzznuc-on-Swoose.html

```shell
~/programs/EMBOSS-6.6.0/emboss/fuzznuc \
-sequence ~/data/P_generosa/genomes/Pgenerosa_v074.fa \
-pattern CG \
-outfile 20190821_fuzznuc_pgenv074.gff \
-rformat gff
```

**Comment by sr320** [comment]
Also note I will separately be dealing with the general ability to count CG at https://github.com/RobertsLab/resources/issues/1207

**Reply by kubu4** [reply]
> normally EMBOSS Fuzznuc....

This is definitely the way to go!

**Comment by yaaminiv** [comment, marked as answer]
Here's my CG motif track (will add to the handbook): http://owl.fish.washington.edu/halfshell/genomic-databank/cgigas_uk_roslin_v1_fuzznuc_CGmotif.gff Similar count to the 13080574 from `fgrep`

Looks the same as Steven's:

![Screen Shot 2021-05-11 at 1 06 03 PM](https://user-images.githubusercontent.com/22335838/117877624-a4b39c80-b259-11eb-9d67-2b77f62e0fef.png)
